Switch latin layouts to use the simple parser

Helium314 opened the pull request:
This PR aims at replacing latin layouts with simplified ones that should look (nearly) the same as currently.
Language-dependent moreKeys are already added.
Layouts are still missing (except for the incomplete qwerty), adding them is the next step and necessary for proper testing and working on details.

Helium314 commented:
This PR will come with some slight changes in layouts:
* Serbian (latin) has slightly narrower keys in bottom row
* German (German) has a symmetric key arrangement instead of being slightly shifted to the right
* layouts with empty space in bottom row (e.g. Danish) have keys better aligned with the row above
* German (German) shifted should show ẞ instead of ß, but shows SS. This is consistent with upcasing of ß to SS in moreKeys, but moreKeys also (always) contains the capitalized ẞ. I think modifying `StringUtils.toTitleCaseOfKeyLabel` to change ß capitalization from SS to ẞ should be fine.
* tablet layouts possibly have different comma moreKeys (not yet tested)
* azerty now always has the `'` key, even when using a manually shifted layout. Previously it would change to `?` when pressing shift (not when auto-shifted). @BlackyHawky do you think this change is ok, or should the current behavior be kept?
(`'` key is missing some moreKeys, will fix)

BlackyHawky commented:
I'm sorry but it's impossible to test because the keyboard is not displayed at all. Here's what appears:

<details>
<summary><b>See screenshot</b></summary>
<br>
<img width=200 src="https://github.com/Helium314/openboard/assets/139015663/a0547ba9-1786-4a72-b649-b3a8a5473f87">
</details>

> azerty now always has the ' key, even when using a manually shifted layout.

However, I don't think it will be disturbing to leave it as you describe. In my case, I always use the symbol view to write the question mark.

> tablet layouts possibly have different comma moreKeys (not yet tested)

I will also test this.

Helium314 commented:
Thiis likely caused by a crash in the keyboard parser or when reading moreKeys.
I can't reproduce it, could you get the stacktrace from logcat?

BlackyHawky commented:
> I can't reproduce it, could you get the stacktrace from logcat?

Stacktrace in attached file
[Stacktrace.txt](https://github.com/Helium314/openboard/files/13391260/Stacktrace.txt)

Helium314 commented:
What I could read is that there is a normal (non-functional) key with an empty label... but where this would come from I don't understand.
I added some logging for finding empty labels using `SimpleParser` as tag, could you give it another try?

BlackyHawky commented:
Unfortunately, the keyboard still won't display. 😬

Attached is a new Stacktrace:
[Stacktrace.txt](https://github.com/Helium314/openboard/files/13392715/Stacktrace.txt)

Helium314 commented:
I found a possibly related issue, maybe that's what caused your crashes. If  it's not fixed, there should be much better logging now.

The issue seems to be a morekey starting with `|`, but with at least one more character. And I have no clue which (normal) key this could be.

BlackyHawky commented:
I'm not sure I can help you...
Here's the latest Stacktrace following your last commit:
[Stacktrace.txt](https://github.com/Helium314/openboard/files/13394932/Stacktrace.txt)

Helium314 commented:
The log confirms my suspicion that the `|` moreKey on e causes the crash. But even in your log it always shows as single character, which cannot trigger the crash as it happens.
In the log I don't find any hidden (non-printable) character after `|`, but maybe they get removed as part of the logging.

I did yet another change to get more logging, i.e. log character codes in that case, which hopefully identifies what's causing the crash.

So far I have only one vague idea: Are you using Windows? Then there might be a way that somehow a carriage return `\r` is inserted before the newline `\n`. No idea how though. Neither git not Android Studio should modify the text... only maybe if you do some manual copy paste.

The crash itself should be fixable easily, but I'd really like to first understand what is going on with the additional character...

Helium314 commented:
If my suspicion is correct, the fix would be in `SimpleLayoutParser`
```diff
-        layoutContent.split("\n\n").mapIndexedTo(mutableListOf()) { i, row ->
+        layoutContent.replace("\r\n", "\n").split("\n\n").mapIndexedTo(mutableListOf()) { i, row ->
             row.split("\n").mapNotNull {
                 if (it.isBlank()) return@mapNotNull null
-                val split = it.split(" ")
+                val split = it.split("\\s+".toRegex())
```
(the second chang might not be necessary in this case, but probably good to have)

BlackyHawky commented:
> So far I have only one vague idea: Are you using Windows?

Indeed, this is the OS I use.

> If my suspicion is correct, the fix would be in SimpleLayoutParser

I made your modifications above and everything works perfectly!!!!
Without it, here is the stacktrace : [Stacktrace.txt](https://github.com/Helium314/openboard/files/13400196/Stacktrace.txt)

However, what exactly do I have to test?
Everything is exactly the same as before.

You said : "_azerty now always has the `'` key, even when using a manually shifted layout. Previously it would change to ? when pressing shift (not when auto-shifted)_ but when I press the shift key, `?` appears as before... 🤔🤔

The only things that don't work are the options:
- "Add variants used in many languages"
- "Add all available variants"

Helium314 commented:
> Without it, here is the stacktrace : Stacktrace.txt

Thanks, indeed for whatever reason a _carriage return_ (`\r`) makes its way into the files... that's pretty awful, but we'll need to deal with things like this.

> The only things that don't work are the options:

The should work, but at the moment the keyboard has to be created again. I do it via restarting the app.

> Everything is exactly the same as before.

Oops, sorry. My "comparison mode" is still enabled, where only the non-shifted layout is the new one (very useful to compare key positions and moreKeys). You can test the full version when changing in `KeyboardBuilder`
```diff
-if (id.mElementId == KeyboardId.ELEMENT_ALPHABET
+if (id.isAlphabetKeyboard
```

BlackyHawky commented:
> You can test the full version when changing in KeyboardBuilder

Now I know the difference! 🤣🤣
Honestly, you can keep the current behavior. It's absolutely not disturbing!
In any case, the keyboard is exactly the same as before. Fantastic work 😲😲

If I understand correctly, the position is set in the code and not in the xml files?

>>The only things that don't work are the options:

>The should work, but at the moment the keyboard has to be created again. I do it via restarting the app.

How exactly do you do it? If I restart from Android Studio or force close from my phone, the keyboard doesn't appear.

Helium314 commented:
> Honestly, you can keep the current behavior. It's absolutely not disturbing!

Great, then it's mostly just fixing a few bugs before merging.

> If I understand correctly, the position is set in the code and not in the xml files?

Yes, the code should be able to determine a good split position, and obviously fails here. Thanks for noticing!

> How exactly do you do it?

Either I click _run app_ in Android Studio, or force stop and then select the keyboard again (because the keyboard app switches on force stop).
Btw this issue also affects the moreKeys when multiple languages are enabled. E.g. when you enable Turkish multilingual typing, you should get some special `g` moreKey, but it only works if Turkish language was already added before the keyboard was created. Just need to find the right place to reset it...

Helium314 commented:
> How exactly do you do it?

Oh, damn. I forgot to add the new files to the commit, so obviously it can only work locally for me. Sorry...

BlackyHawky commented:
> Either I click run app in Android Studio, or force stop and then select the keyboard again (because the keyboard app switches on force stop).

I can't do it; unable to display keyboard. It doesn't matter.

> tablet layouts possibly have different comma moreKeys (not yet tested)

I haven't noticed any difference about comma moreKeys.

About the tablet test:

- For the AZERTY keyboard:
  - The `?` and `!` keys are wider; perhaps this has an influence on the width of the keys `w` `x` `c` `v` `b` `n`
  - `¡` and `¿` are displayed in the `!` and `?` but it's not disturbing.

<details>
<summary><b>See screenshot</b></summary>
<br>
<img width=400 src="https://github.com/Helium314/openboard/assets/139015663/4037f3b5-1b19-485f-a41b-690ebfa31ff9">
</details>


- For the QWERTY keyboard:
  - each line has keys with different widths;
  - `?` + `!` keys are wider ;
  - `¡` and `¿` are also displayed in the `!` and `?` but it's not disturbing.

<details>
<summary><b>See screenshot</b></summary>
<br>
<img width=400 src="https://github.com/Helium314/openboard/assets/139015663/08db1e39-1172-4597-a68d-47c09e846158">
</details>

Helium314 commented:
I removed the azerty layout from the switch for now, there are so many layouts with different keys when shift is enabled (including latin ones like dvorak), so I need to implement it anyway. But not in the simple parser, will do later.

> I haven't noticed any difference about comma moreKeys.

By that point I had added some extra code to remove `!` and `?` from tablet moreKeys, but looking at `morekeys_punctuation` and `morekeys_tablet_punctuation`, position of `'` should be different.

> each line has keys with different widths

Should be better now, I set the key width to 9% like it's in the tablet layout files.

BlackyHawky commented:
> Should be better now

Indeed, it's much better!! 💪
Only the second row has wider keys, but that's not really a problem. 👍

Screenshot of current state:
<img width=400 src="https://github.com/Helium314/openboard/assets/139015663/f61d9221-05b8-4bc9-a9c6-ca54867930bd">
